Estimate order total error: "converting to reserve: unable to accurately represent value as int64"

Estimating a purchase (`POST /ob/estimatetotal`) && making a purchase order (`POST /ob/purchase`) is failing with the following payload and response:

Payload:

```JSON
{
    "shipTo": "Elwood Blues",
    "address": "1060 W Addison",
    "city": "Chicago",
    "state": "Illinois",
    "countryCode": "UNITED_STATES",
    "postalCode": "60613",
    "addressNotes": "",
    "items": [
        {
            "listingHash": "QmTeGATiiTEbNwkLosSRivSApstXLof45jWRjSAVoB1QUv",
            "bigQuantity": "1",
            "options": [
                {
                    "name": "Sizes",
                    "value": "Small"
                },
                {
                    "name": "Color",
                    "value": "Red"
                }
            ],
            "shipping": {
                "name": "Worldwide",
                "service": "Standard"
            },
            "memo": "thanks!",
            "coupons": []
        }
    ],
    "moderator": "QmcdkKM2fWCKzTZdpjEY6abzRbDhRRJvNmqwNKVyiDtGky",
    "paymentCoin": "TLTC"
}
```

Reponse:

```JSON
{
    "success": false,
    "reason": "converting to reserve: unable to accurately represent value as int64"
}
``` 

STDOUT also logs: `15:23:36.614 [ERROR] [core/CalculateOrderTotal] Currency not tracked`

--- a/core/order.go
+++ b/core/order.go
@@ -1271,11 +1271,11 @@ func (n *OpenBazaarNode) getPriceInSatoshi(paymentCoin, currencyCode string, amo
 		}
 	}
 
-	reserveValue, err := originValue.ConvertTo(reserveCurrencyDef, originIntoReserveRate)
+	reserveValue, _, err := originValue.ConvertTo(reserveCurrencyDef, originIntoReserveRate)
 	if err != nil {
 		return big.NewInt(0), fmt.Errorf("converting to reserve: %s", err.Error())
 	}
-	resultValue, err := reserveValue.ConvertTo(paymentCurrencyDef, reserveIntoResultRate)
+	resultValue, _, err := reserveValue.ConvertTo(paymentCurrencyDef, reserveIntoResultRate)
 	if err != nil {
 		return big.NewInt(0), fmt.Errorf("converting from reserve: %s", err.Error())
 	}

--- a/repo/currency.go
+++ b/repo/currency.go
@@ -187,16 +187,17 @@ func (v *CurrencyValue) Normalize() (*CurrencyValue, error) {
 	if err != nil {
 		return nil, err
 	}
-	return v.AdjustDivisibility(localDef.Divisibility)
+	val, _, err := v.AdjustDivisibility(localDef.Divisibility)
+	return val, err
 }
 
 // AdjustDivisibility updates the Currency.Divisibility and adjusts the Amount to match the new
 // value. An error will be returned if the new divisibility is invalid or produces an unreliable
 // result. This is a helper function which is equivalent to ConvertTo using a copy of the
 // CurrencyDefinition using the updated divisibility and an exchangeRatio of 1.0
-func (v *CurrencyValue) AdjustDivisibility(div uint) (*CurrencyValue, error) {
+func (v *CurrencyValue) AdjustDivisibility(div uint) (*CurrencyValue, big.Accuracy, error) {
 	if v.Currency.Divisibility == div {
-		return v, nil
+		return v, 0, nil
 	}
 	defWithNewDivisibility := v.Currency
 	defWithNewDivisibility.Divisibility = div
@@ -206,22 +207,23 @@ func (v *CurrencyValue) AdjustDivisibility(div uint) (*CurrencyValue, error) {
 // ConvertTo will perform the following math given its arguments are valid:
 // v.Amount * exchangeRatio * (final.Currency.Divisibility/v.Currency.Divisibility)
 // where v is the receiver, exchangeRatio is the ratio of (1 final.Currency/v.Currency)
-// v and final must both be Valid() and exchangeRatio must not be zero.
-func (v *CurrencyValue) ConvertTo(final CurrencyDefinition, exchangeRatio float64) (*CurrencyValue, error) {
+// v and final must both be Valid() and exchangeRatio must not be zero. The accuracy
+// indicates if decimal values were trimmed when converting the value back to integer.
+func (v *CurrencyValue) ConvertTo(final CurrencyDefinition, exchangeRatio float64) (*CurrencyValue, big.Accuracy, error) {
 	if err := v.Valid(); err != nil {
-		return nil, fmt.Errorf("cannot convert invalid value: %s", err.Error())
+		return nil, 0, fmt.Errorf("cannot convert invalid value: %s", err.Error())
 	}
 	if err := final.Valid(); err != nil {
-		return nil, fmt.Errorf("cannot convert to invalid currency: %s", err.Error())
+		return nil, 0, fmt.Errorf("cannot convert to invalid currency: %s", err.Error())
 	}
 	if exchangeRatio <= 0 {
-		return nil, ErrCurrencyValueNegativeRate
+		return nil, 0, ErrCurrencyValueNegativeRate
 	}
 
 	amt := new(big.Float).SetInt(v.Amount)
 	exRatio := new(big.Float).SetFloat64(exchangeRatio)
 	if exRatio == nil {
-		return nil, fmt.Errorf("exchange ratio (%f) is invalid", exchangeRatio)
+		return nil, 0, fmt.Errorf("exchange ratio (%f) is invalid", exchangeRatio)
 	}
 	newAmount := new(big.Float).SetPrec(53).Mul(amt, exRatio)
 
@@ -230,13 +232,16 @@ func (v *CurrencyValue) ConvertTo(final CurrencyDefinition, exchangeRatio float6
 		finalMagnitude := math.Pow10(int(final.Divisibility))
 		divisibilityRatio := new(big.Float).SetFloat64(finalMagnitude / initMagnitude)
 		newAmount.Mul(newAmount, divisibilityRatio)
-
-		// confirm no significant figures will be lost in the decimal
-		if !newAmount.IsInt() {
-			return nil, ErrCurrencyValueInsufficientPrecision
-		}
 	}
 
-	result, _ := newAmount.Int(nil)
-	return &CurrencyValue{Amount: result, Currency: final}, nil
+	roundedAmount := new(big.Float)
+	newFloat, _ := newAmount.Float64()
+	if newFloat >= 0 {
+		roundedAmount = big.NewFloat(math.Ceil(newFloat))
+	} else {
+		roundedAmount = big.NewFloat(math.Floor(newFloat))
+	}
+	roundedInt, _ := roundedAmount.Int(nil)
+	roundedAcc := big.Accuracy(roundedAmount.Cmp(newAmount))
+	return &CurrencyValue{Amount: roundedInt, Currency: final}, roundedAcc, nil
 }

--- a/repo/currency.go
+++ b/repo/currency.go
@@ -234,7 +234,7 @@ func (v *CurrencyValue) ConvertTo(final CurrencyDefinition, exchangeRatio float6
 		newAmount.Mul(newAmount, divisibilityRatio)
 	}
 
-	roundedAmount := new(big.Float)
+	var roundedAmount *big.Float
 	newFloat, _ := newAmount.Float64()
 	if newFloat >= 0 {
 		roundedAmount = big.NewFloat(math.Ceil(newFloat))

--- a/repo/currency_test.go
+++ b/repo/currency_test.go
@@ -233,16 +233,16 @@ func TestCurrencyValuesAreEqual(t *testing.T) {
 
 func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 	var (
-		zeroRateErr             = "rate must be greater than zero"
-		undefinedCurrencyErr    = "unknown currency"
-		invalidErr              = "cannot convert invalid value"
-		insufficentPrecisionErr = repo.ErrCurrencyValueInsufficientPrecision.Error()
+		zeroRateErr          = "rate must be greater than zero"
+		undefinedCurrencyErr = "unknown currency"
+		invalidErr           = "cannot convert invalid value"
 
 		examples = []struct {
 			value        *repo.CurrencyValue
 			convertTo    repo.CurrencyDefinition
 			exchangeRate float64
 			expected     *repo.CurrencyValue
+			expectedAcc  int8
 			expectedErr  *string
 		}{
 			{ // errors when definition is nil
@@ -250,6 +250,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 				convertTo:    repo.NilCurrencyDefinition,
 				exchangeRate: 0.99999,
 				expected:     nil,
+				expectedAcc:  0,
 				expectedErr:  &undefinedCurrencyErr,
 			},
 			{ // errors zero rate
@@ -257,6 +258,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 				convertTo:    factory.NewCurrencyDefinition("BCH"),
 				exchangeRate: 0,
 				expected:     nil,
+				expectedAcc:  0,
 				expectedErr:  &zeroRateErr,
 			},
 			{ // errors negative rate
@@ -264,20 +266,23 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 				convertTo:    factory.NewCurrencyDefinition("BCH"),
 				exchangeRate: -0.1,
 				expected:     nil,
+				expectedAcc:  0,
 				expectedErr:  &zeroRateErr,
 			},
 			{ // rounds down
 				value:        factory.MustNewCurrencyValue("1", "BTC"),
 				convertTo:    factory.NewCurrencyDefinition("BCH"),
 				exchangeRate: 0.9,
-				expected:     factory.MustNewCurrencyValue("0", "BCH"),
+				expected:     factory.MustNewCurrencyValue("1", "BCH"),
+				expectedAcc:  1,
 				expectedErr:  nil,
 			},
 			{ // handles negative values
 				value:        factory.MustNewCurrencyValue("-100", "BTC"),
 				convertTo:    factory.NewCurrencyDefinition("BCH"),
 				exchangeRate: 0.123,
-				expected:     factory.MustNewCurrencyValue("-12", "BCH"),
+				expected:     factory.MustNewCurrencyValue("-13", "BCH"),
+				expectedAcc:  -1,
 				expectedErr:  nil,
 			},
 			{ // handles zero
@@ -285,6 +290,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 				convertTo:    factory.NewCurrencyDefinition("BCH"),
 				exchangeRate: 0.99999,
 				expected:     factory.MustNewCurrencyValue("0", "BCH"),
+				expectedAcc:  0,
 				expectedErr:  nil,
 			},
 			{ // handles invalid value
@@ -295,6 +301,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 				convertTo:    factory.NewCurrencyDefinition("BTC"),
 				exchangeRate: 0.5,
 				expected:     nil,
+				expectedAcc:  0,
 				expectedErr:  &invalidErr,
 			},
 			{ // handles conversions between different divisibility
@@ -323,6 +330,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 						CurrencyType: repo.Crypto,
 					},
 				},
+				expectedAcc: 0,
 				expectedErr: nil,
 			},
 			{ // handles conversions between different
@@ -352,6 +360,7 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 						CurrencyType: repo.Fiat,
 					},
 				},
+				expectedAcc: 0,
 				expectedErr: nil,
 			},
 			{ // handles conversions which reduce significant figures
@@ -372,14 +381,23 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 					CurrencyType: repo.Crypto,
 				},
 				exchangeRate: 1,
-				expected:     nil,
-				expectedErr:  &insufficentPrecisionErr,
+				expected: &repo.CurrencyValue{
+					Amount: big.NewInt(76544),
+					Currency: repo.CurrencyDefinition{
+						Name:         "SimpleCoin",
+						Code:         "SPC",
+						Divisibility: 2,
+						CurrencyType: repo.Crypto,
+					},
+				},
+				expectedAcc: 1,
+				expectedErr: nil,
 			},
 		}
 	)
 
 	for _, e := range examples {
-		actual, err := e.value.ConvertTo(e.convertTo, e.exchangeRate)
+		actual, actualAcc, err := e.value.ConvertTo(e.convertTo, e.exchangeRate)
 		if err != nil {
 			if e.expectedErr != nil && !strings.Contains(err.Error(), *e.expectedErr) {
 				t.Errorf("expected value (%s) to error with (%s) but returned: %s", e.value, *e.expectedErr, err.Error())
@@ -396,6 +414,11 @@ func TestCurrencyValuesConvertCorrectly(t *testing.T) {
 			t.Errorf("expected converted value to be %s, but was %s", e.expected, actual)
 			t.Logf("\tfor value: (%s) convertTo: (%s) rate: (%f)", e.value, e.convertTo, e.exchangeRate)
 			continue
+		}
+
+		if expectedAcc := big.Accuracy(e.expectedAcc); actualAcc != expectedAcc {
+			t.Errorf("expected converted accuracy to be %s, but was %s", expectedAcc.String(), actualAcc.String())
+			t.Logf("\tfor value: (%s) convertTo: (%s) rate: (%f)", e.value, e.convertTo, e.exchangeRate)
 		}
 	}
 }
@@ -444,7 +467,7 @@ func TestCurrencyValueAdjustDivisibility(t *testing.T) {
 	subject := factory.MustNewCurrencyValue("123000000", "BTC")
 	subject.Currency.Divisibility = sameDiv
 
-	if newValue, err := subject.AdjustDivisibility(sameDiv); err != nil {
+	if newValue, _, err := subject.AdjustDivisibility(sameDiv); err != nil {
 		t.Fatalf("expected same divisibility to not return an error, but did: %s", err.Error())
 	} else {
 		if !newValue.Currency.Equal(subject.Currency) {
@@ -452,7 +475,7 @@ func TestCurrencyValueAdjustDivisibility(t *testing.T) {
 		}
 	}
 
-	if newValue, err := subject.AdjustDivisibility(2); err != nil {
+	if newValue, _, err := subject.AdjustDivisibility(2); err != nil {
 		t.Fatalf("expected new divisibility to not return an error, but did: %s", err.Error())
 	} else {
 		if newValue.Currency.Equal(subject.Currency) {

